conduit/pipeline/pipeline.go

```diff
@@ -431,8 +467,250 @@ func addMetrics(block data.BlockData, importTime time.Duration) {
+
+				selectStart := time.Now()
+				select {
+				case <-p.ctx.Done():
+					return
+				case blkOutChan <- pluginBlk:
+					waitTime := time.Since(selectStart)
+					totalFeedWait += waitTime
+					p.logger.Tracef("processor[%d] %s for round %d into blkOutChan after waiting %dms", idx, proc.Metadata().Name, lastRnd, totalFeedWait.Milliseconds())
+				}
+			}
+		}
+	}()
+}
+
+// exporterHandler handles the exporter's Receive method, updating the metadata's NextRound,
+// saving the metadata, and invoking the activated plugin callbacks. If the context's cancellation is received
+// before the exporter's Receive method has succeeded, cancel immediately.
+// However, after the exporter's Receive() method has succeeded we try every component at least once
+// but will abort if a failure occurs after a cancellation is received.
+func (p *pipelineImpl) exporterHandler(exporter exporters.Exporter, blkChan pluginChannel) {
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		var lastError error
+		lastRound := uint64(0)
+		totalSelectWait := time.Duration(0)
+		totalExportWork := time.Duration(0)
+
+		defer func() {
+			p.logger.Debugf("exporter %s handler exiting. lastRnd=%d totalSelectWait=%dms, totalExportWork=%dms", exporter.Metadata().Name, lastRound, totalSelectWait.Milliseconds(), totalExportWork.Milliseconds())
+			if lastError != nil {
+				err := fmt.Errorf("exporter %s handler (%w) after round %d: %w", exporter.Metadata().Name, errExporterCause, lastRound, lastError)
+				p.cancelWithProblem(err)
+				p.logger.Error(err)
+			}
+		}()
+
+		for {
+			p.logger.Tracef("exporter handler waiting for block data. round=%d", lastRound)
+			selectStart := time.Now()
+			select {
+			case <-p.ctx.Done():
+				p.logger.Infof("exporter handler exiting lastRnd=%d", lastRound)
+				return
+			case pluginBlk := <-blkChan:
+				waitTime := time.Since(selectStart)
+				totalSelectWait += waitTime
+				p.logger.Tracef("exporter handler received block data for round %d after wait of %s", lastRound, waitTime)
+				lastRound = pluginBlk.Round()
+
+				if p.pipelineMetadata.NextRound != lastRound {
+					lastError = fmt.Errorf("aborting after out of order block data. %d != %d", p.pipelineMetadata.NextRound, lastRound)
+					return
+				}
+
+				var exportTime time.Duration
+				exportTime, lastError = RetriesNoOutput(exporter.Receive, pluginBlk.BlockData, p, exporter.Metadata().Name)
+				if lastError != nil {
+					lastError = fmt.Errorf("aborting after failing to export round %d: %w", lastRound, lastError)
+					return
+				}
+				metrics.ExporterTimeSeconds.Observe(exportTime.Seconds())
+				// Ignore round 0 (which is empty).
+				if lastRound > 0 {
+					// Previously we reported time starting after block fetching is complete
+					// through the end of the export operation. Now that each plugin is running
+					// in its own goroutine, report only the time of the export.
+					addMetrics(pluginBlk.BlockData, exportTime)
```

Comment:
why not add an import timestamp to BlockData? then you could have 
`addMetrics(blockdata, time.Since(blockdata.importTime))`

Reply:
After our previous discussion, I temporarily added **importTime** to `blockdata` . Then I decided I didn't like polluting block chain crucial data with performance metrics and also imposing a requirement on downstream processes to know about this pollution (imagine an exporter that just serializes all of the `BlockData` and then gets this extra info it wasn't expecting). There's a lively discussion related to this [on a thread above](https://github.com/algorand/conduit/pull/128#discussion_r1284363465).